src/posts/2020-07-27-unsucking-your-project-onboarding.md

```diff
@@ -0,0 +1,789 @@
+
+We run `set +e` before adding the plugins in case it fails since they're
+probably already added. Afterwards we'll need to import the `nodejs` release
+team keyring, before safely installing node, which is conveniently available on
+a `bin/` helper:
+
+```shell
+#!/usr/bin/env sh
+
+set -ue
+
+## Keys from https://github.com/nodejs/node/#release-team
+KEYS="94AE36675C464D64BAFA68DD7434390BDBE9B9C5 \
+      FD3A5288F042B6850C66B31F09FE44734EB7990E \
+      71DCFD284A79C3B38668286BC97EC7A07EDE3FC1 \
+      DD8F2338BAE7501E3DD5AC78C273792F7D83545D \
+      C4F0DFFF4E8C1A8236409D08E73BC641CC11F4C8 \
+      B9AE9905FFD7803F25714661B63B535A4C206CA9 \
+      8FCCA13FEF1D0C2E91008E09770F7A9A5AE15600 \
+      77984A986EBC2AA786BC0F66B01FBB92821C587A \
+      A48C2BEE680E841632CD4E44F07496B3EB3C1762 \
+      4ED778F539E3634C779C87C6D7062848A1AB005C \
+      B9E2F5981AA6E0CD28160D9FF13993A75599653C"
+
+SERVERS="ha.pool.sks-keyservers.net
+         p80.pool.sks-keyservers.net:80 \
+         ipv4.pool.sks-keyservers.net \
+         keyserver.ubuntu.com
+         keyserver.ubuntu.com:80 \
+         pgp.mit.edu
+         pgp.mit.edu:80"
+
+for key in $KEYS; do
+  if ! gpg --list-keys $key 2>&1 >/dev/null; then
+    for server in $SERVERS; do
+      gpg --no-tty --keyserver "hkp://$server" --display-charset utf-8 --recv-keys "$key"  && break
+    done
+  fi
+done
+```
+
+#### Environment setup
+
+Having the languages available, we then move on to setting up the environment
+through environment variables. This usually happens through `dotenv` (less
+common) or (more common nowadays) `direnv`.
+
+Lately all projects I've been on have been built with `direnv` in mind, so let's
+focus on that:
+
+```shell
+pp_info "setup" "Setting up the env..."
+
+if [ ! -f .envrc ]; then
+  cp .envrc.sample .envrc
+  printf "env file created, you might want to open .envrc and set the correct values. We recommend using direnv to manage it.\n\n"
+else
+  printf "envrc file already exists, skipping...\n\n"
+fi
+
+. "./.envrc"
+```
+
+We always add a `.envrc.sample` to the project to document the environment
+variables we need. The condition makes sure no `.envrc` is present when doing
+the copy as to not override it. **We always try to make sure `bin/setup` is
+idempotent**, allowing others to run it repeatedly without making any damages.
+
+There have also been some variations of this snippet that check for the
+installation of `direnv`, similarly to what we do with `asdf`, but it has since
+been dropped.
+
+As a sidenote, if you are using `dotenv`, replace `.envrc` with `.env` and it
+should work!
+
+#### Installing dependencies
+
+In a typical Elixir + React project, this breaks down into three steps: making
+sure `yarn` is available, installing Node dependencies, installing Elixir
+dependencies:
+
+```shell
+pp_info "setup" "Ensuring yarn is installed"
+if not_installed "yarn"; then
+  curl -o- -L https://yarnpkg.com/install.sh | bash
+else
+  echo "All done!"
+fi
```

Comment:
You actually don't need to check for `yarn`, or install it from a remote script. Yarn is available as an NPM package. So you can instead just install Yarn globally (since we're using asdf-vm it is contained to the Node.js version we're using). If it is already installed, if will just be skipped, otherwise NPM will install it or update it for you.

```suggestion
pp_info "setup" "Ensuring yarn is installed"
npm install -g yarn
```

Reply:
I'm sorry I don't get the argument. If you change the project's Node version, don't you re-run `bin/setup`? Also, if you look at the install script the "agnostic setup" does, that's the tarball of the project decompressed into `~/.yarn`. You're polluting the user's home directory instead of installing a tool in the contained environment set by asdf-vm.

I have nothing against someone preferring the "agnostic setup", but I would be utterly pissed off if I started working on a project and suddenly it started installing random stuff in my home folder without asking.

Reply:
And installing globally is fine then?

Reply:
"globally" here applies to the Node.js version you're running only. If you change it, it goes away. If you delete it, it goes away without influencing any other version. Any other tool I would even argue we don't need to install it globally, because we can simply install it as a project dependency and run it as `npx tool`. But since Yarn installs the dependencies, it has to be "global" just like NPM (which with asdf-vm is also Node.js version dependent).

I might be reading too much into your message but I'm starting to get a vibe that you're feeling attacked and I really had no intention of it. Merely wanted to be as helpful as I could. Please proceed as you see fit. :pray:

Reply:
But npm is bound to node versions, yarn isn't. It's supposed to act as an independent tool, hence they recommending that option.

Reply:
But that's not the option they recommend, it's precisely to install Yarn globally with NPM. https://yarnpkg.com/getting-started/install#global-install